File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ListNode* first = head;
        ListNode* second;

        while (first->next != NULL) {
            second = first->next;
            int a = first->val;
            int b = second->val;
            ListNode* tmp = new ListNode(gcd(a, b));
            tmp->next = second;
            first->next = tmp;
            first = second;
        }
        
        return head;
    }
};

File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:

        def gcd(a, b):
            while b > 0:
                a, b = b, a % b
            return a
        
        first = head
        while first.next != None:
            second = first.next
            tmp = ListNode()
            tmp.val = gcd(first.val, second.val)
            tmp.next = second
            first.next = tmp
            first = second

        return head